### PR TITLE
e2e tests for `useOverflow` logic

### PR DESCRIPTION
**Pull request by r100-stack**
## Changes

Pulled out of #2120 as PR 1 in the PR stack ([PR stack order](https://github.com/iTwin/iTwinUI/pull/2120#issuecomment-2386574990)).

To make that PR easy to review, according to @mayank99's suggestion, I created this PR to only add new e2e tests that test the existing `useOverflow` logic. These tests will thus act as a baseline that the new `useOverflow` logic will also have to pass to make sure that there is no change in the resulting behavior.

Some e2e tests (e.g. `MiddleTextTrucation`) seem to be failing only on GitHub (e.g. [action run](https://github.com/iTwin/iTwinUI/actions/runs/9996764547/job/27631858596?pr=2151#step:7:339)) but not locally. I guess this is a small testing environment specific issue that I am currently trying to fix. Thus, the PR is ready for review despite the failing tests.

TODO:
* [x] Fix the flaky/unreliable e2e tests.

## Testing

* Ensured that the new e2e tests are passing.

## Docs

N/A

## After PR TODOs

- [ ] Consider a lint rule for `await`ing the `expect`s in e2e tests ([#2151 (comment)](https://github.com/iTwin/iTwinUI/pull/2151#discussion_r1684484281))
- [ ] Don't rely on implementation details when it's possible to pass props to the tag container/tags in ComboBox ([#2151 (comment)](https://github.com/iTwin/iTwinUI/pull/2151#discussion_r1684394649)) and Select ([#2151 (comment)](https://github.com/iTwin/iTwinUI/pull/2151#discussion_r1684395240)).

### CODE REVIEW

**Review by mayank99 — COMMENTED** [review]
Are you also planning to delete the existing unit tests? Because otherwise it begs the question: why even add the e2e tests?

File: testing/e2e/app/routes/Breadcrumbs/spec.ts
Comment (mayank99) [inline testing/e2e/app/routes/Breadcrumbs/spec.ts:33]: nit: "overflow button",  not "overflow tag"

File: testing/e2e/app/routes/Breadcrumbs/spec.ts
Comment (mayank99) [inline testing/e2e/app/routes/Breadcrumbs/spec.ts:57]: optional: one layer of abstraction could be removed by simply creating a `setContainerSize` that accepts both the `page` object and the actual value.

same comment applies to `expectOverflowState` and any other similar utils in other test scenarios

File: testing/e2e/app/routes/Breadcrumbs/spec.ts
Comment (mayank99) [inline testing/e2e/app/routes/Breadcrumbs/spec.ts:67]: nit: `dimension` makes it sound like the possible values would be `horizontal` or `vertical`. It might be better to name this `size` or even just `value`

File: testing/e2e/app/routes/ButtonGroup/route.tsx
Comment (mayank99) [inline testing/e2e/app/routes/ButtonGroup/route.tsx:4]: nit: always use `* as React` namespace import, not default import

same comment applies to all other instances

File: testing/e2e/app/routes/ButtonGroup/route.tsx
Comment (mayank99) [inline testing/e2e/app/routes/ButtonGroup/route.tsx:20]: components should never be created in a nested way like this, or they will be unnecessarily recreated on every render and might cause some issues with React. If data needs to be shared from the parent component, use props.

File: testing/e2e/app/routes/DropdownMenu/spec.ts
Comment (mayank99) [inline testing/e2e/app/routes/DropdownMenu/spec.ts:1]: this file shouldn't have changed?

File: testing/e2e/app/routes/ComboBox/route.tsx
Comment (mayank99) [inline testing/e2e/app/routes/ComboBox/route.tsx:40]: all the code for new tests should be placed in a separate component and conditionally rendered by the default component

in other words: adding new tests should not affect existing tests where possible. right now, you need to change the existing tests to account for the extra button

File: testing/e2e/app/routes/ComboBox/spec.ts
Comment (mayank99) [inline testing/e2e/app/routes/ComboBox/spec.ts:14]: nit: do not nest `describe` blocks; instead create sibling `describe` blocks at the top-level. The existing `describe` block could be renamed to something like "ComboBox (general)" if you want.

same comment applies to other tests that use this pattern

File: testing/e2e/app/routes/MiddleTextTruncation/spec.ts
Comment (mayank99) [inline testing/e2e/app/routes/MiddleTextTruncation/spec.ts:26]: don't forget to `await`

File: testing/e2e/app/routes/Select/route.tsx
Comment (mayank99) [inline testing/e2e/app/routes/Select/route.tsx:36]: nit: never write nested ternaries; regular `if` conditions are always more readable

File: testing/e2e/app/routes/ComboBox/spec.ts
Comment (mayank99) [inline testing/e2e/app/routes/ComboBox/spec.ts:26]: this locator is relying on an implementation detail which could change at any time?

File: testing/e2e/app/routes/Select/spec.ts
Comment (mayank99) [inline testing/e2e/app/routes/Select/spec.ts:131]: this locator is also relying on an implementation detail?

**Review by mayank99 — APPROVED** [review]
Some minor comments to improve tests and reduce diff.

Approving on the assumption that you can make CI pass.

File: testing/e2e/app/routes/ButtonGroup/route.tsx
Comment (mayank99) [reply testing/e2e/app/routes/ButtonGroup/route.tsx]: we have had this rule for a while now, it's in the style guide

the default `import React` is considered legacy at this point, it's from before hooks and before the new JSX transform was introduced. i believe the React team wants to remove it in a future version, so they encourage namespace (`import * as React`) or named imports (`import {} from`)

**Review by Ben-Pusey-Bentley — APPROVED** [review]
Approving so that this isn't held up. Added some comments that I hope will help with the failing e2e tests on the CI.

File: testing/e2e/app/routes/ComboBox/spec.ts
Comment (mayank99) [reply testing/e2e/app/routes/ComboBox/spec.ts]: The `select-tag` is still coming from an internal class which can change at any time.

Right now we don't have a way to pass classes/attributes to the tag container or the tags (#894). This test could probably be rewritten to avoid relying on such implementation details (e.g. by only comparing the text inside the main container), but I'll leave it upto you if you want to rewrite it or add a TODO to wait until we allow customizing the tag container.

File: testing/e2e/app/routes/Select/spec.ts
Comment (mayank99) [reply testing/e2e/app/routes/Select/spec.ts]: Same as https://github.com/iTwin/iTwinUI/pull/2151#discussion_r1684394649, this is still relying on an implementation detail.

File: testing/e2e/app/routes/ButtonGroup/route.tsx
Comment (mayank99) [reply testing/e2e/app/routes/ButtonGroup/route.tsx]: i noticed it in some other tests too (e.g. Table)

File: testing/e2e/app/routes/MiddleTextTruncation/spec.ts
Comment (mayank99) [reply testing/e2e/app/routes/MiddleTextTruncation/spec.ts]: i noticed a few more `expect`s that were missing `await`s.

we might need a lint rule to help avoid this situation

File: testing/e2e/app/routes/ButtonGroup/route.tsx
Comment (mayank99) [reply testing/e2e/app/routes/ButtonGroup/route.tsx]: this one is a pretty gnarly mistake imo, it should be fixed before merging

Comment (mayank99) [reply testing/e2e/app/routes/ButtonGroup/route.tsx]: you can copy-paste this if you'd like:

<details><summary>route.tsx</summary>

```tsx
import {
  Table,
  tableFilters,
  TablePaginator,
  TablePaginatorRendererProps,
} from '@itwin/itwinui-react';
import { useSearchParams } from '@remix-run/react';
import * as React from 'react';

export default function TableTest() {
  const { exampleType } = useConfig();

  return exampleType === 'withTablePaginator' ? (
    <WithTablePaginator />
  ) : (
    <Default />
  );
}

function useConfig() {
  const [searchParams] = useSearchParams();

  const exampleType = searchParams.get('exampleType') || 'default';
  const disableResizing = searchParams.get('disableResizing') === 'true';
  const columnResizeMode = searchParams.get('columnResizeMode') || 'fit';
  const maxWidths = searchParams.getAll('maxWidth');
  const minWidths = searchParams.getAll('minWidth');
  const density = (searchParams.get('density') || undefined) as
    | 'default'
    | 'condensed'
    | 'extra-condensed'
    | undefined;
  const isSelectable = searchParams.get('isSelectable') === 'true';
  const subRows = searchParams.get('subRows') === 'true';
  const filter = searchParams.get('filter') === 'true';
  const selectSubRows = !(searchParams.get('selectSubRows') === 'false');
  const enableVirtualization = searchParams.get('virtualization') === 'true';
  const empty = searchParams.get('empty') === 'true';
  const scroll = searchParams.get('scroll') === 'true';
  const oneRow = searchParams.get('oneRow') === 'true';
  const stateReducer = searchParams.get('stateReducer') === 'true';
  const scrollRow = Number(searchParams.get('scrollRow'));

  return {
    exampleType,
    disableResizing,
    columnResizeMode,
    maxWidths,
    minWidths,
    density,
    isSelectable,
    subRows,
    filter,
    selectSubRows,
    enableVirtualization,
    empty,
    scroll,
    oneRow,
    stateReducer,
    scrollRow,
  };
}

function Default() {
  const {
    disableResizing,
    columnResizeMode,
    maxWidths,
    minWidths,
    density,
    isSelectable,
    subRows,
    filter,
    selectSubRows,
    enableVirtualization,
    empty,
    scroll,
    oneRow,
    stateReducer,
    scrollRow,
  } = useConfig();

  const virtualizedData = React.useMemo(() => {
    const size = oneRow ? 1 : 100000;
    const arr = new Array(size);
    if (!empty) {
      for (let i = 0; i < size; ++i) {
        arr[i] = {
          index: i,
          name: `Name${i}`,
          description: `Description${i}`,
          id: i,
        };
      }
    }
    return arr;
  }, [oneRow, empty]);

  const data = subRows
    ? [
        {
          index: 1,
          name: 'Name1',
          description: 'Description1',
          id: '111',
        },
        {
          index: 2,
          name: 'Name2',
          description: 'Description2',
          subRows: [
            {
              index: 2.1,
              name: 'Name2.1',
              description: 'Description2.1',
              id: '223',
            },
            {
              index: 2.2,
              name: 'Name2.2',
              description: 'Description2.2',
              id: '224',
            },
          ],
          id: '222',
        },
        {
          index: 3,
          name: 'Name3',
          description: 'Description3',
          subRows: [
            {
              index: 3.1,
              name: 'Name3.1',
              description: 'Description3.1',
              id: '334',
            },
            {
              index: 3.2,
              name: 'Name3.2',
              description: 'Description3.2',
              id: '335',
            },
          ],
          id: '333',
        },
      ]
    : [
        {
          index: 1,
          name: 'Name1',
          description: 'Description1',
          id: '111',
        },
        {
          index: 2,
          name: 'Name2',
          description: 'Description2',
          id: '222',
        },
        {
          index: 3,
          name: 'Name3',
          description: 'Description3',
          id: '333',
        },
      ];

  const isRowDisabled = React.useCallback(
    (rowData: Record<string, unknown>) => {
      return rowData.name === 'Name3.2';
    },
    [],
  );

  return (
    <>
      <Table
        columns={[
          {
            Header: '#',
            accessor: 'index',
            width: 100,
            maxWidth: parseInt(maxWidths[0]) || undefined,
            minWidth: parseInt(minWidths[0]) || undefined,
          },
          {
            Header: 'Name',
            accessor: 'name',
            maxWidth: parseInt(maxWidths[1]) || undefined,
            minWidth: parseInt(minWidths[1]) || undefined,
            disableResizing,
            Filter: filter ? tableFilters.TextFilter() : undefined,
          },
          {
            Header: 'Description',
            accessor: 'description',
            width: '200px',
          },
          {
            Header: 'ID',
            accessor: 'id',
            width: '8rem',
          },
        ]}
        data={enableVirtualization ? virtualizedData : data}
        emptyTableContent='No data.'
        isResizable
        isRowDisabled={isRowDisabled}
        isSelectable={isSelectable}
        isSortable
        density={density}
        columnResizeMode={columnResizeMode as 'fit' | 'expand' | undefined}
        selectSubRows={selectSubRows}
        enableVirtualization={enableVirtualization}
        style={enableVirtualization ? { maxHeight: '90vh' } : undefined}
        scrollToRow={
          scroll
            ? (rows, data) =>
                rows.findIndex((row) => row.original === data[scrollRow])
            : undefined
        }
        stateReducer={
          stateReducer
            ? (newState, action, previousState, instance) => {
                if (action.type === 'toggleRowSelected') {
                  console.log(action.value);
                }
                return newState;
              }
            : undefined
        }
      />
    </>
  );
}

function WithTablePaginator() {
  const { density } = useConfig();

  const columns = React.useMemo(
    () => [
      {
        id: 'name',
        Header: 'Name',
        accessor: 'name',
        Filter: tableFilters.TextFilter(),
      },
      {
        id: 'description',
        Header: 'Description',
        accessor: 'description',
        maxWidth: 200,
        Filter: tableFilters.TextFilter(),
      },
    ],
    [],
  );

  const data = React.useMemo(
    () =>
      Array(505)
        .fill(null)
        .map((_, index) => ({
          name: `Name ${index}`,
          description: `Description ${index}`,
        })),
    [],
  );

  const paginator = React.useCallback(
    (props: TablePaginatorRendererProps) => (
      <TablePaginator id='paginator' {...props} />
    ),
    [],
  );

  return (
    <>
      <div id='container' style={{ height: '80vh' }}>
        <Table
          style={{ height: '100%' }}
          emptyTableContent='No data.'
          columns={columns}
          data={data}
          pageSize={50}
          density={density}
          paginatorRenderer={paginator}
        />
      </div>
    </>
  );
}
```

</details>

**Review by mayank99 — APPROVED** [review]

**Pull request merged**
